### PR TITLE
Batched torch.eig() and use of magma for CUDA torch.eig() operation

### PR DESCRIPTION
**Pull request by mfkasim1**
This is a new pull request because I messed up my previous branch.
This pull request is to add the capability of `torch.eig` in handling batched tensor.

### CODE REVIEW

File: aten/src/ATen/native/LinearAlgebraUtils.h [hunk 1]
```diff
@@ -94,6 +94,10 @@ static inline void batchCheckErrors(std::vector<int64_t>& infos, const char* nam
     } else if (info > 0) {
       if (strstr(name, "svd")) {
         AT_ERROR(name, ": the updating process of SBDSDC did not converge (error: ", info, ")");
+      } else if (strstr(name, "eig")) {
+        AT_ERROR(name, ": For batch ", i, ": the QR algorithm failed to compute all the eigenvalues, "
+                 "and no eigenvectors have been computed; elements and ", info+1, ":N of W contain "
```
Comment (albanD) [inline aten/src/ATen/native/LinearAlgebraUtils.h]: Given that this is raising an error, will the content actually be properly filled with the right values? If it's an out-of-place operation, the user won't even have access to the result.
I think you can either make this a warning with TORCH_WARN stating that some elements are wrong. Or make this an error only stating that the QR algorithm failed for that batch element.
What do you thinkg?

File: aten/src/ATen/native/cuda/BatchLinearAlgebra.cu [hunk 2]
```diff
@@ -1183,6 +1209,95 @@ std::tuple<Tensor,Tensor> _qr_helper_cuda(const Tensor& self, bool some) {
...
+  magma_int_t lwork = -1;
+  scalar_t wkopt;
+  scalar_t* eigvecs_data = eigenvectors ? eigvecs.data_ptr<scalar_t>() : NULL;
+  magmaEig<scalar_t>(jobvl, jobvr, n, self_data, n, reigvals_data, ieigvals_data, NULL, 1, eigvecs_data, ldvr, &wkopt, lwork, &info);
+
+  scalar_t* work;
+  lwork = magma_int_cast(wkopt, "work_size");
+  ALLOCATE_ARRAY(work, scalar_t, lwork);
+
+  for (int64_t i = 0; i < batch_size; i++) {
```
Comment (albanD) [inline aten/src/ATen/native/cuda/BatchLinearAlgebra.cu]: Why not start this at `1` given that you computed the 0th batch above already?

File: aten/src/ATen/native/native_functions.yaml [hunk 3]
```diff
@@ -4701,16 +4701,16 @@
     CPU: _symeig_helper_cpu
     CUDA: _symeig_helper_cuda
 
-- func: eig.e(Tensor self, bool eigenvectors=False, *, Tensor(a!) e, Tensor(b!) v) -> (Tensor(a!) eigenvalues, Tensor(b!) eigenvectors)
-  dispatch:
-    CPU: legacy::cpu::_th_eig_out
-    CUDA: legacy::cuda::_th_eig_out
+- func: eig.e(Tensor self, bool eigenvectors=False, *, Tensor(a!) e, Tensor(b!) V) -> (Tensor(a!) eigenvalues, Tensor(b!) eigenvectors)
```
Comment (albanD) [inline aten/src/ATen/native/native_functions.yaml]: cc @gchanan do you know why this overload is not `.out` ?

**Review by albanD — COMMENTED** [review]
This looks good.
Some surprising things but I guess you followed very closely symeig right?

Let me know if you need help with the xla tests.

**Comment by albanD** [comment]
cc @nikitaved would something like this be useful as part of the linalg work? Do we want to revive it?

**Comment by nikitaved** [comment]
@albanD , is this operation not deprecated in favor of `torch.linalg.eig`?

**Comment by albanD** [comment]
It is. But would something similar to this be beneficial to linalg.eig?

**Comment by nikitaved** [comment]
We are slowly removing deprecated linear operations in the `torch` namespace from the codebase, so merging it could result in double work. I assume it is safe to abandon `torch.eig`, as the one from `linalg` is better both from the forward as well as from the backward :)

**Comment by albanD** [comment]
I do not mean that we should merge this. Or add any new feature to torch.eig no. I am asking if similar batch support for linalg.eig is something we want? And if so, can we re-use some of the scaffolding in this PR (blas bindings)?

Like, should we completely forget about this PR or keep it in the back of our mind when we add batched support as it will help with some of the implementation details?

**Comment by nikitaved** [comment]
All linalg methods support batches. Some of them go an extra mile with the CUDA support, i.e. magma implementation is reserved for large inputs, while medium to small sizes are delegated to cuBLAS and cuSolver.

**Comment by albanD** [comment]
Ok! Thanks for the details here.
We can re-open this later if we have specific use for it.

**Pull request closed without merging**